### PR TITLE
Apply patches when used in radare2

### PR DESCRIPTION
**Pull request by satk0**
I've made a PR that fixes all the issues we found in r2 as I said I would do in issue #559

Based on the following commits:
- https://github.com/radareorg/radare2/pull/23315/commits/96c44c644e79540fa5acbb54c584a16996f36ae9
- https://github.com/radareorg/radare2/pull/23315/commits/68464192a3cc215fb07f503e6ae65241f794c0fb
- https://github.com/radareorg/radare2/commit/86529d7cf63fa399b7e2f59680c553e9dd3edfd3
- https://github.com/radareorg/radare2/commit/b2a467cd13c561042554901313ebdcb749eb0de7
- https://github.com/radareorg/radare2/pull/23315/commits/6fd44351782abb555ac163a62b0f2fc85ac780da

### CODE REVIEW

File: cutils.c [hunk 1]
```diff
@@ -1129,7 +1129,7 @@ void rqsort(void *base, size_t nmemb, size_t size, cmp_f cmp, void *opaque)
 
 /*---- Portable time functions ----*/
 
-#if defined(_MSC_VER)
+#if defined(_MSC_VER) || __MINGW32__
```
Comment (saghul) [inline cutils.c]: Why is this necessary? Is it to support the old mingw32 aka not mingw-w64?

File: quickjs.c [hunk 2]
```diff
@@ -62,7 +62,7 @@
 #define NO_TM_GMTOFF
 #endif
 
-#if !defined(EMSCRIPTEN) && !defined(__wasi__)
+#if !defined(EMSCRIPTEN) && !defined(__wasi__) && defined(__STDC_NO_ATOMICS__) && __STDC_NO_ATOMICS__
```
Comment (saghul) [inline quickjs.c:65]: Shouldn't this be negated? We want the check to succeed if __STDC_NO_ATOMICS__ is not defined, right?

File: quickjs.c [hunk 3]
```diff
@@ -34561,7 +34561,7 @@ static JSString *JS_ReadString(BCReaderState *s)
     }
 #ifdef DUMP_READ_OBJECT
     if (check_dump_flag(s->ctx->rt, DUMP_READ_OBJECT)) {
-        bc_read_trace(s, "");  // hex dump and indentation
+        bc_read_trace(s, "%s", ""); // hex dump and indentation
```
Comment (saghul) [inline quickjs.c]: Ha, nice! Can you please remove the silencer flag I added to CMake then?

**Review by saghul — COMMENTED** [review]
Thank you! I left a couple of comments.

File: cutils.c (hunk 1 above)
Comment (satk0) [reply cutils.c]: Probably yea, that's @trufae's commit, here:
https://github.com/radareorg/radare2/commit/86529d7cf63fa399b7e2f59680c553e9dd3edfd3
 and sheesh its failing all the mingw checks, idk why :/

File: quickjs.c (hunk 2 above)
Comment (satk0) [reply quickjs.c:65]: Hmmm, yea you are totally right, I've messed it up XD

Comment (satk0) [reply quickjs.c:65]: Idk how it passed the ci jobs for r2, magic XDD

File: quickjs.c (hunk 3 above)
Comment (satk0) [reply quickjs.c]: sure

File: cutils.c (hunk 1 above)
Comment (saghul) [reply cutils.c]: Mingw32 is really old and no longer maintained AFAIK, mingw-w64 (which also has a 32bit version, despite the name) has been the de facto MinGW for a while I think.

@trufae ?

**Comment by satk0** [comment]
Done, maybe this time mingw-w64 will pass, I added a check for only mingw32, excluding w64 :pray:

**Comment by satk0** [comment]
Btw, thanks a lot for being so responsive, it is really motivating!

File: libbf.c [hunk 4]
```diff
@@ -1710,6 +1710,11 @@ static int __bf_div(bf_t *r, const bf_t *a, const bf_t *b, limb_t prec,
         slimb_t d;
 
         na = n + nb;
+
+        if (na >= (SIZE_MAX / sizeof(limb_t)) - 1) {
```
Comment (saghul) [inline libbf.c:1715]: This fails on Windows, since IIRC the limb bits are fewer there.

D:/a/quickjs/quickjs/libbf.c:1714:16: error: comparison is always false due to limited range of data type [-Werror=type-limits]

File: cutils.c [hunk 5]
```diff
@@ -1129,7 +1129,7 @@ void rqsort(void *base, size_t nmemb, size_t size, cmp_f cmp, void *opaque)
 
 /*---- Portable time functions ----*/
 
-#if defined(_MSC_VER)
+#if defined(_MSC_VER) || (__MINGW32__ && !defined(__MINGW64__))
```
Comment (saghul) [inline cutils.c]: Come to think of it, we could use this everywhere on windows, perhaps that's a simpler thing, let's try it please.

File: libbf.c (hunk 4 above)
Comment (satk0) [reply libbf.c:1715]: Huh, that's rough, will try to play with it more

File: cutils.c (hunk 5 above)
Comment (satk0) [reply cutils.c]: Yea I need to find a macro that distinguish mingw32 from Mingw-w64

Comment (saghul) [reply cutils.c]: I was thinking on the opposite direction, as in, to use this implementation also in MinGW, if possible.

Comment (satk0) [reply cutils.c]: Ooh, allright, I could try

Comment (satk0) [reply cutils.c]: Btw can you plz give me any hint about the limb bits on windows?

Comment (saghul) [reply cutils.c]: See https://github.com/quickjs-ng/quickjs/blob/1eb9608d64242d0efec199b24f3bdf13063bf88e/libbf.h#L34 the size of limb_t is platform dependent, so perhaps an extra if-def is necessary to check.

Comment (satk0) [reply cutils.c]: Thanks a lot, I will try this today as well as your `__MINGW32__` idea

Comment (saghul) [reply cutils.c]: 👍

**Comment by satk0** [comment]
Done, letme know if that's what you expected. Sorry for pushing it so late :sweat:

File: libbf.c [hunk 6]
```diff
@@ -1710,6 +1710,11 @@ static int __bf_div(bf_t *r, const bf_t *a, const bf_t *b, limb_t prec,
         slimb_t d;
 
         na = n + nb;
+
+        if (na >= (SIZE_MAX / sizeof(limb_t)) - 1) && !defined(_WIN32) {
```
Comment (saghul) [inline libbf.c]: I think this check should be surrounded by some ifdef sizeof (limb_t) > sizeof(uint64_t) instead

File: cutils.c (hunk 1 above)
Comment (saghul) [reply cutils.c]: Can either of you confirm which version of MinGW you're using?

My fear is we'd break mingw32 support since we don't run tests on it.

Comment (satk0) [reply cutils.c]: IMG_20241005_142544

Comment (saghul) [reply cutils.c]: That's mingw-w64 which we already support, I don't think those changes are necessary.

Any chance you can try to build radare without them?

File: libbf.c (hunk 6 above)
Comment (satk0) [reply libbf.c]: W8, I haven't closed the brackets here, that's why its all failing, sorry. I will fix it

**Comment by satk0** [comment]
Chck now, sry for this obv mistake I've made

**Comment by satk0** [comment]
Allright, some checks are still failing, I think that changing this comparison back to a simple `!defined(_WIN32)` would solve it.

**Comment by saghul** [comment]
> Allright, some checks are still failing, I think that changing this comparison back to a simple `!defined(_WIN32)` would solve it.

I think so too!

**Comment by satk0** [comment]
Ok, letsee how it goes :exclamation:

**Comment by saghul** [comment]
Ah I thought you meant something else.

I think what you want to check for then is if LIMB_LOG2_BITS is 6 or not.

**Comment by satk0** [comment]
Oh, ok so I'd check for if `LIMB_LOG2_BITS == 6` instead of `_WIN32`

Thanks for the quick review btw 😃

**Comment by satk0** [comment]
Another thing is that I should check if `defined(__MINGW32__)` instead of only if `__MINGW32__`

**Comment by saghul** [comment]
> Another thing is that I should check if `defined(__MINGW32__)` instead of only if `__MINGW32__`

I'd rather see it gone. I think it's not necessary since both us and radars are using mingw-w64.

**Comment by satk0** [comment]
Ok, So I contacted pancake on it and he said this:
IMG_20241006_203533

So yea, we use mingw32 in r2,  I just asked pancake on mingw version, not mingw32 (I wasn't precise enough). So he answered with the mingw-w64 one XD.

But nevertheless, I'll go with your suggestion to omit this and see what happens with mingw32 on ubuntu24, thanks!

**Comment by saghul** [comment]
Sounds good!

**Comment by satk0** [comment]
OK done, hopefully nothing will break 🙏

**Review by saghul — APPROVED** [review]

**Comment by saghul** [comment]
Looks good! Let's see if the CI is green!

**Comment by satk0** [comment]
Hmm, seems I have to put this limb thing inside `#if LIMB_LOG2_BITS == 6` cuz C is still calculating it cuz operator precedence 😖

**Comment by satk0** [comment]
Ok done, chck it now

**Comment by saghul** [comment]
Squash - rebased them and landed! https://github.com/quickjs-ng/quickjs/pull/582

**Comment by saghul** [comment]
Thanks for persevering! ❤️
